### ISSUE TITLE
Plugin checker developer documentation

### ISSUE THREAD

**Issue by eclarke1** [10up/plugin-check#83]
## Developer Documentation

As part of the Milestone 1 review an initial draft of the developer documentation could be written to help highlight any potential improvements to the overall architecture.

The overall aim of writing the documentation is to discover ways to improve the Plugin Check API to make it simpler and easier to create checks.

The documentation will be written in a `/docs` folder and written in markdown.

The developer documentation should cover a number of aspects of extending the plugin checker, in particular creating and including a custom check.

### Basic Outline

- **General Introduction**
    - Overview of how the plugin checker works from a technical perspective.
    - What are static and runtime checks and their differences.
    - Highlight important details such as:
        - How runtime checks only run against active plugins.
        - How runtime checks may require preparations and what they are.
    - Describe how the test environment is created and prepared to not impact the actual site.
    - How to extend the plugin checker. Not only contributing checks to the main repository, but also creating custom checks to be used outside of the main repo (by using the `wp_plugin_check_checks` filter).
- **Creating a Static Check**
    - Extending the `Abstract_PHP_CodeSniffer_Check` (Discuss both the `Check` and `Static_Check` interfaces)
    - Creating a `get_args()` method and returning an array of args
       - `'sniffs' => 'SNIFF_NAME (E.g. WordPress.WP.I18n)'`.
       - `'standard' => 'STANDARD' (E.g., WordPress, WordPressVIPMinimum, and so on.'`
    - Adding the check to the Plugin Checker via the `wp_plugin_check_checks` filter with a key and instance. 
- **Creating a Runtime Check**
    - Implementing the `Abstract_Runtime_Check`
    - Discuss the `Runtime_Check` and `Preparation` interfaces.
    - **Different types of Preparations**
        - Check Preparations - Created with the `prepare()` method on the Check class.
        - Preparation Classes - Classes implementing the `Preparation` interface and contains a `prepare()` method.
        - Shared Preparations - Checks that implement the `With_Shared_Preparations` interface with a `get_shared_preparations()` method.
    - The Checks `run()` method used to perform the check.
        - Discuss the `Check_Result` instance passed.
            - Adding errors and warnings with the `add_message()` method.
        - Using the `Plugin_Context` instance to get the plugins path/urls to perform checks.
    - Adding the check to the Plugin Checker via the `wp_plugin_check_checks` filter with a key and instance.

**Comment by eclarke1** [comment]
Assigning to @jjgrainger to please define a structured plan here, ready to work on as part of Sprint 6 (Feb 27 - Mar 10). Aiming to have documentation written by March 10 so in Sprint 7 (Mar 13 - 24) we can use it to better complete QA testing and architectural code review.

**Comment by jjgrainger** [comment]
@eclarke1 I am going to assign this to @vishalkakadiya to pick up with his remaining hours on the project.

@vishalkakadiya can you create a bullet point outline of what the top-level pages/headings we would want to write documentation for as part of milestone 1 review. We don't need to go into detail, just highlight the key areas we want to cover.

This will include items such as, creating a check (both static and runtime), creating check preparations (at a check level, shared preparation), running checks with the CLI command and WP Admin page etc.

If you have any questions just ask and I can support here. Thanks!

**Comment by vishalkakadiya** [comment]
@jjgrainger I have added very basic details of the developer documentation here. Let me know your feedback on it, Thanks! 🙂

**Comment by jjgrainger** [comment]
Hi @felixarntz I've updated the issue description here with a basic outline of what the initial developer documentation should cover as part of the milestone 1 review. Please can you review and let me know if you think there is anything missing. Thanks.

**Comment by felixarntz** [comment]
@jjgrainger Are the top level points in the issue description supposed to be individual articles each, or parts of a single article? I assume individual articles, but wanted to check.

This list looks solid, a few points of feedback:
* I think it would be great to have one more general introduction article for developers/contributors, something like "How the plugin checker works". Not to be an actual user guide, but more how it works from a technical perspective, so that contributors understand the concepts, e.g. what are static checks, what are runtime checks, quirks like runtime checks only running if the plugin to test is active, how the plugin checker runs runtime checks without "messing up" the actual site, etc. That could be the first article to write.
* For both the articles about how to write static checks and how to write runtime checks, it would be good to clarify somewhere that they can be useful both for contributing to the plugin checker and also for people that just want to write their own checks for their custom use (the latter is specifically possible because there is a filter, so you can hook in your checks from another separate plugin for example).
* I think the article on writing test cases is lower priority. It's great you added this and I think writing tests is something that is often missed from documentation. But at the same time I wonder whether we should rather make this part of the articles on how to write checks, potentially as a V2. In other words, we could have a separate effort once the initial articles are done to expand those two with a new section on how to write test cases for the check classes.

**Comment by jjgrainger** [comment]
Thanks @felixarntz 

Yes, the top level points are supposed to be individual articles, with the sub points highlighting their content.

I've included a general introduction here covering the points you've raised. I've also included your second bullet point about contributing, both to the plugin and custom projects in this section too.

I agree with your third point on writing test cases being part of a future effort. For that I have created a new issue here for us to update at a later date - https://github.com/10up/plugin-check/issues/107

Let me know if theres anything more needed here otherwise we can move this to todo. Thanks!
